### ISSUE TITLE
Parse error when using function overloads

### ISSUE THREAD

**Issue by matt-oakes** [typescript-eslint/typescript-eslint#481]
**What code were you trying to parse?**

```ts
export function overloadExample(
  numberOrString: string,
): number;
export function overloadExample(
  numberOrString: number,
): object;
export function overloadExample(
  numberOrString: number | string,
): number | object {
  if (typeof numberOrString === 'number') {
    return {}
  } else {
    return 1;
  }
}
```

This code makes use of [function overloads](http://www.typescriptlang.org/docs/handbook/functions.html#overloads). The error occurs with any function overload, not just this specific one.

**What did you expect to happen?**
Code to parse correctly.

**What actually happened?**
```
$ eslint 'js/**/*.{js,ts,tsx}' 'example/**/*.{js,ts,tsx}'
SyntaxError: Unexpected token ; (3:10)
   5 | export function overloadExample(
   6 |   numberOrString: string,
>  7 | ): number;
     |          ^
   8 | export function overloadExample(
   9 |   numberOrString: number,
  10 | ): object;
Occurred while linting /Users/matt/projects/OS/react-native-netinfo/js/test.ts:5
    at t (/Users/matt/projects/OS/react-native-netinfo/node_modules/prettier/parser-flow.js:1:310)
    at Object.parse (/Users/matt/projects/OS/react-native-netinfo/node_modules/prettier/parser-flow.js:1:1429899)
    at Object.parse$2 [as parse] (/Users/matt/projects/OS/react-native-netinfo/node_modules/prettier/index.js:10633:19)
    at coreFormat (/Users/matt/projects/OS/react-native-netinfo/node_modules/prettier/index.js:13842:23)
    at format (/Users/matt/projects/OS/react-native-netinfo/node_modules/prettier/index.js:14101:73)
    at formatWithCursor (/Users/matt/projects/OS/react-native-netinfo/node_modules/prettier/index.js:14117:12)
    at /Users/matt/projects/OS/react-native-netinfo/node_modules/prettier/index.js:42518:15
    at Object.format (/Users/matt/projects/OS/react-native-netinfo/node_modules/prettier/index.js:42537:12)
    at Program (/Users/matt/projects/OS/react-native-netinfo/node_modules/eslint-plugin-prettier/eslint-plugin-prettier.js:415:45)
    at listeners.(anonymous function).forEach.listener (/Users/matt/projects/OS/react-native-netinfo/node_modules/eslint/lib/util/safe-emitter.js:45:58)
```

**Versions**

| package                     | version |
| --------------------------- | ------- |
| `@typescript-eslint/parser` | `1.7.0` |
| `TypeScript`                | `3.4.5` |
| `ESLint`                    | `5.16.0` |
| `node`                      | `10.13.0` |
| `npm`                       | `6.4.1` |

**Comment by bradzacher** [comment]
Could you please post your eslint config?

The call stack says `prettier/parser-flow.js` which is called by `eslint-plugin-prettier`

What is your prettier config as well?

**Comment by matt-oakes** [comment]
Thanks @bradzacher!

`.eslintrc.js`:

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  parserOptions: {
    ecmaVersion: 6,
    sourceType: "module",
    project: "./tsconfig.json"
  },
  extends: ["@react-native-community"],
  plugins: ["@typescript-eslint"],
  overrides: [
    {
      files: ["**/*.ts"],
      rules: {
        // These rules are from the @typescript-eslint recommended config, with the ones which conflict with prettier removed
        "@typescript-eslint/adjacent-overload-signatures": "error",
        "@typescript-eslint/array-type": "error",
        "@typescript-eslint/ban-types": "error",
        "camelcase": "off",
        "@typescript-eslint/camelcase": "error",
        "@typescript-eslint/class-name-casing": "error",
        "@typescript-eslint/explicit-function-return-type": "warn",
        "@typescript-eslint/explicit-member-accessibility": "error",
        "indent": "off",
        "@typescript-eslint/interface-name-prefix": "error",
        "@typescript-eslint/no-angle-bracket-type-assertion": "error",
        "no-array-constructor": "off",
        "@typescript-eslint/no-array-constructor": "error",
        "@typescript-eslint/no-empty-interface": "error",
        "@typescript-eslint/no-explicit-any": "warn",
        "@typescript-eslint/no-inferrable-types": "error",
        "@typescript-eslint/no-misused-new": "error",
        "@typescript-eslint/no-namespace": "error",
        "@typescript-eslint/no-non-null-assertion": "error",
        "@typescript-eslint/no-object-literal-type-assertion": "error",
        "@typescript-eslint/no-parameter-properties": "error",
        "@typescript-eslint/no-triple-slash-reference": "error",
        "no-unused-vars": "off",
        "@typescript-eslint/no-unused-vars": "warn",
        "@typescript-eslint/no-use-before-define": "error",
        "@typescript-eslint/no-var-requires": "error",
        "@typescript-eslint/prefer-interface": "error",
        "@typescript-eslint/prefer-namespace-keyword": "error",
      }
    },
    {
      files: ["**/__tests__/**/*.ts", "**/*.spec.ts"],
      env: {
        jest: true
      },
      rules: {
        "@typescript-eslint/explicit-function-return-type": "off"
      }
    }
  ],
};
```

The `@react-native-community` config is here:

https://github.com/facebook/react-native/blob/master/packages/eslint-config-react-native-community/index.js

`.prettierrc`:

```json
{
  "requirePragma": true,
  "singleQuote": true,
  "trailingComma": "all",
  "bracketSpacing": false,
  "jsxBracketSameLine": true,
  "parser": "flow"
}
```

**Comment by bradzacher** [comment]
There's your problem!
You've hard configured prettier to use the flow parser, which is why it's breaking on your typescript code. Flow has no concept of body-less function definitions.

Could you please try deleting that line and linting agian?

**Comment by matt-oakes** [comment]
Deleting the line you mentioned in `.prettierrc` to the code below gives the same error.

```json
{
  "requirePragma": true,
  "singleQuote": true,
  "trailingComma": "all",
  "bracketSpacing": false,
  "jsxBracketSameLine": true
}
```

If I simplify my `.eslintrc.js` to this gives the same error as well (turns out my duplicate configuration was not required):

```javascript
module.exports = {
  extends: ["@react-native-community"],
  overrides: [
    {
      files: ["**/__tests__/**/*.ts", "**/*.spec.ts"],
      env: {
        jest: true
      },
      rules: {
        "@typescript-eslint/explicit-function-return-type": "off"
      }
    }
  ],
};
```

Removing the `extends: ["@react-native-community"]` line from my original ESLint config gets rid of the error, so I assume it's a configuration issue within that config. I know this isn't your job to look into, but can see see anything in [the config](https://github.com/facebook/react-native/blob/master/packages/eslint-config-react-native-community/index.js) which could be causing this issue?

**Comment by matt-oakes** [comment]
@bradzacher I have found the issue! It is this line of the React Native config:

https://github.com/facebook/react-native/blob/master/packages/eslint-config-react-native-community/index.js#L219

The `fb` part implicitly sets the parser to flow, overriding the value in `.prettierrc`.

Thanks for your help in figuring this out!

**Comment by bradzacher** [comment]
prettier should default to the correct parser if you don't overwrite it, so that makes sense.
good work finding that.

It's good to know what to do for future!

**Comment by matt-oakes** [comment]
That config was extracted from the React Native project which always had its own special rules with prettier, so it's a bit of a mess. I'm going to try to fix it up now 👍 

Thanks again for the help and the really useful tool 😄
